### PR TITLE
Make parameter info handler asynchronous

### LINKED ISSUE #3900: Please do not display "Preparing parameter info..." popup window
Or let me know how to turn this feature off.
It's extremely annoying, it eats my keypresses and doesn't let me use your wonderful plugin.

toolchain 1.36.0
plugin 0.2.98.2124-191
idea IC-191.7141.44
macOS

### PR DESCRIPTION
**Pull request by vlad20012**
Another hack to make "parameter info handler"s asynchronous. Also, gets rid of "Preparing parameter info..." popup

F'ixes #3900

### CODE REVIEW

**Comment by mvlabat** [comment]
That's an awesome change, @vlad20012, thank you!

File: src/test/kotlin/org/rust/ide/hints/RsParameterInfoHandlerTestBase.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+package org.rust.ide.hints
+
+import com.intellij.lang.parameterInfo.ParameterInfoHandler
+import com.intellij.psi.PsiElement
+import com.intellij.testFramework.utils.parameterInfo.MockCreateParameterInfoContext
+import com.intellij.testFramework.utils.parameterInfo.MockParameterInfoUIContext
+import com.intellij.testFramework.utils.parameterInfo.MockUpdateParameterInfoContext
+import org.rust.RsTestBase
+
+abstract class RsParameterInfoHandlerTestBase<A: PsiElement, B>(
```
Comment (Undin) [inline src/test/kotlin/org/rust/ide/hints/RsParameterInfoHandlerTestBase.kt]:
```suggestion
abstract class RsParameterInfoHandlerTestBase<A : PsiElement, B>(
```

Comment (Undin) [reply src/test/kotlin/org/rust/ide/hints/RsParameterInfoHandlerTestBase.kt]: 👍

File: src/main/kotlin/org/rust/ide/hints/RsAsyncParameterInfoHandler.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import org.rust.lang.core.psi.ext.startOffset
+import org.rust.openapiext.isUnitTestMode
+import java.util.concurrent.Callable
+import java.util.concurrent.Executor
+
+/**
+ * This is a hack to make [ParameterInfoHandler] asynchronous. Usual [ParameterInfoHandler] is called from the EDT
+ * and so complex computations inside it (e.g. name resolution) can freeze the UI.
+ */
+abstract class RsAsyncParameterInfoHandler<ParameterOwner: PsiElement, ParameterType: Any>
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/hints/RsAsyncParameterInfoHandler.kt]:
```suggestion
abstract class RsAsyncParameterInfoHandler<ParameterOwner : PsiElement, ParameterType : Any>
```

File: src/main/kotlin/org/rust/ide/hints/RsAsyncParameterInfoHandler.kt [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    : ParameterInfoHandler<ParameterOwner, ParameterType> {
+
+    abstract fun findTargetElement(file: PsiFile, offset: Int): ParameterOwner?
+
+    abstract fun calculateParameterInfoAsync(element: ParameterOwner): Array<ParameterType>?
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/ide/hints/RsAsyncParameterInfoHandler.kt]: Any suggestions? :(

File: src/main/kotlin/org/rust/ide/hints/RsParameterInfoHandler.kt [hunk 4]
```diff
@@ -6,124 +6,66 @@
...
             context.defaultParameterColor)
     }
-
-    private fun findElementForParameterInfo(contextElement: PsiElement) =
-        contextElement.ancestorStrict<RsValueArgumentList>()
-
-    /**
-     * Finds index of the argument in the given place
-     */
-    private fun findArgumentIndex(place: PsiElement): Int {
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/ide/hints/RsParameterInfoHandler.kt]: +

**Comment by vlad20012** [comment]
Fixed

**Pull request merged**